File: apps/web/src/delivery.ts
import type { BrowserAccess } from './api';
import type { CollectionExecution } from './collections';
import { isHex } from './graphs';

export interface DeliveryInput { runId: string; taskId: string; artifactDigest: string; baseBranch: string; title: string; description: string }
export interface ProviderCheck { id: string; name: string; commit: string; state: string }
export interface ProviderDeployment { id: string; repositoryId: string; providerProfile: string; commit: string; commitRelation: string; environment: string; environmentId?: string; productionEnvironment?: boolean; state: string; providerState: string; statusId?: string; statusTruncated: boolean; providerUpdatedAt: string; observedAt: string }
export interface DeliveryObservation { sequence: number; providerId: string; number: number; url: string; commit: string; tree: string; state: string; draft: boolean; mergeCommit?: string; checks: ProviderCheck[]; checksTruncated: boolean; checksState: string; deployment: string; deployments: ProviderDeployment[]; deploymentsTruncated: boolean; productionOutcome: string; observedAt: string; trigger?: {kind: string; key?: string; eventType?: string; payloadDigest?: string} }
export interface Delivery { id: string; workspaceId: string; repositoryId: string; proposerId: string; input: DeliveryInput; digest: string; target: {workspaceId: string; repositoryId: string; provider: string; host: string; providerId: string; locator: string; profile: string; integrationVersion: number}; baseCommit: string; baseTree: string; resultTree: string; patchDigest: string; branch: string; createdAt: string; authorization?: {actor: string; digest: string; createdAt: string}; receipt?: {digest: string; observation: DeliveryObservation; createdAt: string}; observation?: DeliveryObservation; execution?: CollectionExecution }
export interface DeliveryPage { deliveries: {id: string; workspaceId: string; repositoryId: string; digest: string; title: string; authorized: boolean; state: string; createdAt: string}[]; nextBefore?: string }
export interface ExecutionEvidence { id: string; repositoryId: string; argv: string[]; state: string; exitCode?: number; output?: string; outputDigest: string; truncated: boolean; startedAt: string; finishedAt: string; sourceDigest: string }
export interface ExecutionPatch { repositoryId: string; baseCommit: string; baseTree: string; resultTree: string; patch: string; digest: string; paths: string[] }
export interface ExecutionArtifact { cleanupConfirmed: boolean; inputDigest: string; profileDigest: string; image: string; adapter: string; adapterVersion: string; producer: ExecutionEvidence; patches: ExecutionPatch[]; checks: ExecutionEvidence[]; startedAt: string; finishedAt: string }
export interface DeliveryArtifact { deliveryId: string; deliveryDigest: string; artifactDigest: string; artifact: ExecutionArtifact }
export interface InspectedExecutionArtifact { artifact: ExecutionArtifact; decodedPatches: {metadata: ExecutionPatch; bytes: Uint8Array; text: string}[] }
export interface InspectedArtifact extends DeliveryArtifact, InspectedExecutionArtifact {}
export const deliveryPath = (id: string) => `/repository-deliveries/${encodeURIComponent(id)}`;
const date = (v: unknown) => typeof v === 'string' && Number.isFinite(Date.parse(v));
export function validateDelivery(value: Delivery, access: BrowserAccess, id?: string) {
  if(!value||!isHex(value.id,32)||id!==undefined&&value.id!==id||value.workspaceId!==access.workspaceID||value.repositoryId!==access.repositoryID||!isHex(value.digest,64)||!value.input||!isHex(value.input.runId,32)||!isHex(value.input.taskId,32)||!isHex(value.input.artifactDigest,64)||typeof value.input.title!=='string'||typeof value.input.description!=='string'||typeof value.input.baseBranch!=='string'||!value.target||value.target.workspaceId!==access.workspaceID||value.target.repositoryId!==access.repositoryID||!['github','gitlab'].includes(value.target.provider)||value.target.host!==value.target.provider+'.com'||!isHex(value.baseCommit,40)||!isHex(value.baseTree,40)||!isHex(value.resultTree,40)||!isHex(value.patchDigest,64)||typeof value.branch!=='string'||!date(value.createdAt))throw new Error('The delivery does not match the selected scope or exact artifact.');
  if(value.authorization!==undefined&&(!value.authorization||value.authorization.digest!==value.digest||typeof value.authorization.actor!=='string'||!date(value.authorization.createdAt)))throw new Error('The publication authorization is incomplete.');
  for(const o of [value.receipt?.observation,value.observation])if(o){o.checks=o.checks??[];o.deployments=o.deployments??[];if(!isHex(o.commit,40)||o.tree!==value.resultTree||typeof o.state!=='string'||typeof o.checksState!=='string'||!Array.isArray(o.checks)||o.checks.length>200||o.checks.some(c=>!c||c.commit!==o.commit||typeof c.name!=='string'||typeof c.state!=='string')||typeof o.checksTruncated!=='boolean'||!Array.isArray(o.deployments)||o.deployments.length>100||o.deployments.some(d=>!d||d.repositoryId!==value.repositoryId||d.providerProfile!==value.target.profile||!isHex(d.commit,40)||typeof d.environment!=='string'||typeof d.state!=='string'||!date(d.observedAt))||typeof o.deploymentsTruncated!=='boolean'||!date(o.observedAt))throw new Error('Provider observations are incomplete or do not match the delivery.');}
}
export function validateDeliveryPage(page: DeliveryPage, access: BrowserAccess) {
  if(!page||!Array.isArray(page.deliveries)||page.deliveries.length>20||page.deliveries.some(d=>!d||!isHex(d.id,32)||!isHex(d.digest,64)||d.workspaceId!==access.workspaceID||d.repositoryId!==access.repositoryID||typeof d.title!=='string'||typeof d.authorized!=='boolean'||typeof d.state!=='string'||!date(d.createdAt))||new Set(page.deliveries.map(d=>d.id)).size!==page.deliveries.length||page.nextBefore!==undefined&&(typeof page.nextBefore!=='string'||page.nextBefore.length>1024))throw new Error('The delivery list is incomplete. Refresh before inspecting.');
}
export function providerURL(url: string, host: string): string | undefined { try{const value=new URL(url);if(value.protocol==='https:'&&value.hostname===host&&!value.username&&!value.password&&!value.port)return value.href;}catch{/* Display an unavailable link for invalid provider data. */} }
const digestBytes = async (bytes: Uint8Array) => Array.from(new Uint8Array(await crypto.subtle.digest('SHA-256',bytes as BufferSource)),b=>b.toString(16).padStart(2,'0')).join('');
export { visibleControls } from './sourceText';

// The server verifies the typed immutable artifact digest under all run grants.
// Also verify every displayed patch/check byte digest locally before enabling
// publication. The larger artifact response is never silently sliced for review.
export async function inspectArtifact(value: DeliveryArtifact, selected: Delivery): Promise<InspectedArtifact> {
  if(!value||value.deliveryId!==selected.id||value.deliveryDigest!==selected.digest||value.artifactDigest!==selected.input.artifactDigest||!value.artifact)throw new Error('The artifact does not match the inspected publication proposal.');
  const inspected=await inspectExecutionArtifact(value.artifact);
  const {artifact:a,decodedPatches:decoded}=inspected;
  const p=decoded.find(p=>p.metadata.repositoryId===selected.repositoryId)?.metadata;
  if(!p||p.digest!==selected.patchDigest||p.baseCommit!==selected.baseCommit||p.baseTree!==selected.baseTree||p.resultTree!==selected.resultTree||new Set(a.patches.map(p=>p.repositoryId)).size!==a.patches.length)throw new Error('The artifact lacks the exact target repository patch.');
  return {...value,...inspected};
}

// Failed attempts and read-only design reports use the same complete byte
// validation as delivery, without publication eligibility or a required patch.
export async function inspectExecutionArtifact(value: ExecutionArtifact): Promise<InspectedExecutionArtifact> {
  const a=value;
  a.patches=a.patches??[];a.checks=a.checks??[];
  if(!isHex(a.inputDigest,64)||!isHex(a.profileDigest,64)||typeof a.image!=='string'||typeof a.adapter!=='string'||typeof a.adapterVersion!=='string'||typeof a.cleanupConfirmed!=='boolean'||!a.producer||typeof a.producer.state!=='string'||!Array.isArray(a.patches)||a.patches.length>16||!Array.isArray(a.checks)||a.checks.length>32||!date(a.startedAt)||!date(a.finishedAt))throw new Error('Implementation evidence is incomplete.');
  const decoded:InspectedArtifact['decodedPatches']=[];
  for(const p of a.patches){
    if(!p||typeof p.repositoryId!=='string'||!isHex(p.baseCommit,40)||!isHex(p.baseTree,40)||!isHex(p.resultTree,40)||!isHex(p.digest,64)||!Array.isArray(p.paths)||p.paths.length>10000||p.paths.some(path=>typeof path!=='string')||typeof p.patch!=='string'||p.patch.length>12*1024*1024)throw new Error('A patch is incomplete or exceeds its bounds.');
    const bytes=Uint8Array.from(atob(p.patch),c=>c.charCodeAt(0));if(bytes.length>8*1024*1024||await digestBytes(bytes)!==p.digest)throw new Error('A displayed patch does not match its retained byte digest.');
    decoded.push({metadata:p,bytes,text:new TextDecoder('utf-8',{fatal:true}).decode(bytes)});
  }
  for(const c of [a.producer,...a.checks]){if(!c||typeof c.id!=='string'||typeof c.repositoryId!=='string'||!Array.isArray(c.argv??[])||(c.argv??[]).some(arg=>typeof arg!=='string')||typeof c.state!=='string'||typeof c.truncated!=='boolean'||!isHex(c.outputDigest,64)||!isHex(c.sourceDigest,64)||c.output!==undefined&&typeof c.output!=='string'||await digestBytes(new TextEncoder().encode(c.output??''))!==c.outputDigest)throw new Error('A check does not match its retained output digest.');}
  return {artifact:a,decodedPatches:decoded};
}

export function publicationEvidenceReady(a: InspectedArtifact, repositoryID: string) { return a.artifact.cleanupConfirmed && [a.artifact.producer,...a.artifact.checks].every(c=>c.state==='passed'&&c.exitCode===0&&!c.truncated) && a.artifact.checks.some(c=>c.repositoryId===repositoryID); }

File: internal/domain/delivery.go
package domain

import (
	"encoding/json"
	"strings"
	"time"
	"unicode"
	"unicode/utf8"
)

const (
	GitHubDeliveryProfile = "github-delivery/2026-03-10"
	GitLabDeliveryProfile = "gitlab-delivery/v4-19.3"
)

// DeliveryInput names immutable implementation evidence, never caller-supplied
// patches. A separate human authorization binds the resulting proposal digest.
type DeliveryInput struct {
	RunID          string `json:"runId"`
	TaskID         string `json:"taskId"`
	ArtifactDigest string `json:"artifactDigest"`
	BaseBranch     string `json:"baseBranch"`
	Title          string `json:"title"`
	Description    string `json:"description"`
}
type DeliveryAuthorization struct {
	Actor     string    `json:"actor"`
	Digest    string    `json:"digest"`
	CreatedAt time.Time `json:"createdAt"`
}
type DeliveryIntegrationConfig struct {
	WorkspaceID  string   `json:"workspaceId"`
	RepositoryID string   `json:"repositoryId"`
	Profile      string   `json:"profile"`
	Locator      string   `json:"locator"`
	CredentialID string   `json:"credentialId"`
	BaseBranches []string `json:"baseBranches"`
	Enabled      bool     `json:"enabled"`
}
type DeliveryTarget struct {
	WorkspaceID        string `json:"workspaceId"`
	RepositoryID       string `json:"repositoryId"`
	Provider           string `json:"provider"`
	Host               string `json:"host"`
	ProviderID         string `json:"providerId"`
	Locator            string `json:"locator"`
	Profile            string `json:"profile"`
	IntegrationVersion int64  `json:"integrationVersion"`
}

// ProviderCheck reports an observation at an exact provider commit. No check
// rows, neutral or skipped results, and unavailable APIs cannot imply passing.
type ProviderCheck struct {
	ID     string `json:"id"`
	Name   string `json:"name"`
	Commit string `json:"commit"`
	State  string `json:"state"`
}
type ProviderDeployment struct {
	ID                    string    `json:"id"`
	RepositoryID          string    `json:"repositoryId"`
	ProviderProfile       string    `json:"providerProfile"`
	Commit                string    `json:"commit"`
	CommitRelation        string    `json:"commitRelation"`
	Environment           string    `json:"environment"`
	EnvironmentID         string    `json:"environmentId,omitempty"`
	ProductionEnvironment *bool     `json:"productionEnvironment,omitempty"`
	State                 string    `json:"state"`
	ProviderState         string    `json:"providerState"`
	StatusID              string    `json:"statusId,omitempty"`
	StatusTruncated       bool      `json:"statusTruncated"`
	ProviderUpdatedAt     time.Time `json:"providerUpdatedAt"`
	ObservedAt            time.Time `json:"observedAt"`
}
type DeliveryObservationTrigger struct {
	Kind          string `json:"kind"`
	Key           string `json:"key,omitempty"`
	EventType     string `json:"eventType,omitempty"`
	PayloadDigest string `json:"payloadDigest,omitempty"`
}
type DeliveryObservation struct {
	Deployments          []ProviderDeployment        `json:"deployments"`
	DeploymentsTruncated bool                        `json:"deploymentsTruncated"`
	Trigger              *DeliveryObservationTrigger `json:"trigger,omitempty"`
	Sequence             int64                       `json:"sequence"`
	ProviderID           string                      `json:"providerId"`
	Number               int64                       `json:"number"`
	URL                  string                      `json:"url"`
	Commit               string                      `json:"commit"`
	Tree                 string                      `json:"tree"`
	State                string                      `json:"state"`
	Draft                bool                        `json:"draft"`
	MergeCommit          string                      `json:"mergeCommit,omitempty"`
	Checks               []ProviderCheck             `json:"checks"`
	ChecksTruncated      bool                        `json:"checksTruncated"`
	ChecksState          string                      `json:"checksState"`
	Deployment           string                      `json:"deployment"`
	ProductionOutcome    string                      `json:"productionOutcome"`
	ObservedAt           time.Time                   `json:"observedAt"`
}
type DeliveryReceipt struct {
	Digest      string              `json:"digest"`
	Observation DeliveryObservation `json:"observation"`
	CreatedAt   time.Time           `json:"createdAt"`
}
type Delivery struct {
	Receipt       *DeliveryReceipt       `json:"receipt,omitempty"`
	ID            string                 `json:"id"`
	WorkspaceID   string                 `json:"workspaceId"`
	RepositoryID  string                 `json:"repositoryId"`
	ProposerID    string                 `json:"proposerId"`
	Input         DeliveryInput          `json:"input"`
	Digest        string                 `json:"digest"`
	Target        DeliveryTarget         `json:"target"`
	BaseCommit    string                 `json:"baseCommit"`
	BaseTree      string                 `json:"baseTree"`
	ResultTree    string                 `json:"resultTree"`
	PatchDigest   string                 `json:"patchDigest"`
	Branch        string                 `json:"branch"`
	CreatedAt     time.Time              `json:"createdAt"`
	Authorization *DeliveryAuthorization `json:"authorization,omitempty"`
	Observation   *DeliveryObservation   `json:"observation,omitempty"`
	Execution     *CollectionExecution   `json:"execution,omitempty"`
}
type DeliverySummary struct {
	ID             string    `json:"id"`
	WorkspaceID    string    `json:"workspaceId"`
	RepositoryID   string    `json:"repositoryId"`
	RunID          string    `json:"runId"`
	TaskID         string    `json:"taskId"`
	ArtifactDigest string    `json:"artifactDigest"`
	Digest         string    `json:"digest"`
	Title          string    `json:"title"`
	Branch         string    `json:"branch"`
	CreatedAt      time.Time `json:"createdAt"`
	Authorized     bool      `json:"authorized"`
	ReceiptDigest  string    `json:"receiptDigest,omitempty"`
	State          string    `json:"state"`
}
type DeliveryPage struct {
	Deliveries []DeliverySummary `json:"deliveries"`
	NextBefore string            `json:"nextBefore,omitempty"`
}

func ValidDeliveryBranch(value string) bool {
	if value == "" || len(value) > 128 || !utf8.ValidString(value) || strings.TrimSpace(value) != value || strings.HasPrefix(value, "-") || strings.HasPrefix(value, "/") || strings.HasSuffix(value, "/") || strings.HasSuffix(value, ".") || strings.Contains(value, "..") || strings.Contains(value, "@{") || strings.ContainsAny(value, " ~^:?*[\\") || strings.ContainsFunc(value, unicode.IsControl) {
		return false
	}
	for _, part := range strings.Split(value, "/") {
		if part == "" || strings.HasPrefix(part, ".") || strings.HasSuffix(part, ".lock") {
			return false
		}
	}
	return true
}
func ValidateDeliveryInput(input DeliveryInput) error {
	if !IsLowerHex(input.RunID, 32) || !IsLowerHex(input.TaskID, 32) || !IsLowerHex(input.ArtifactDigest, 64) || !ValidDeliveryBranch(input.BaseBranch) || !validAccessText(input.Title, 240) || len(input.Description) > 16<<10 || !utf8.ValidString(input.Description) || strings.ContainsRune(input.Description, 0) {
		return ErrInvalidInput
	}
	return nil
}
func ValidateDeliveryIntegration(config DeliveryIntegrationConfig) error {
	if ValidateAccessID(config.WorkspaceID) != nil || ValidateAccessID(config.RepositoryID) != nil || ValidateAccessID(config.CredentialID) != nil || len(config.BaseBranches) < 1 || len(config.BaseBranches) > 8 {
		return ErrInvalidInput
	}
	branches := map[string]bool{}
	for _, branch := range config.BaseBranches {
		if !ValidDeliveryBranch(branch) || branches[branch] {
			return ErrInvalidInput
		}
		branches[branch] = true
	}
	switch config.Profile {
	case GitHubDeliveryProfile:
		parts := strings.Split(config.Locator, "/")
		if len(parts) != 2 {
			return ErrInvalidInput
		}
		for _, part := range parts {
			if part == "" || part == "." || part == ".." || len(part) > 100 || strings.ContainsFunc(part, func(r rune) bool {
				return !(r >= 'a' && r <= 'z' || r >= 'A' && r <= 'Z' || r >= '0' && r <= '9' || r == '.' || r == '-' || r == '_')
			}) {
				return ErrInvalidInput
			}
		}
	case GitLabDeliveryProfile:
		if config.Locator != "" {
			return ErrInvalidInput
		}
	default:
		return ErrInvalidInput
	}
	return nil
}

// ValidateDeliveryObservation protects the durable evidence boundary. Provider
// state may establish a merge, but never an unobserved deployment or outcome.
func ValidateDeliveryObservation(d Delivery, o DeliveryObservation) error {
	if o.Number < 1 || o.ProviderID == "" || len(o.ProviderID) > 128 || !IsLowerHex(o.Commit, 40) || o.Tree != d.ResultTree || o.ObservedAt.IsZero() || len(o.Checks) > 200 || len(o.Deployments) > 100 || o.ProductionOutcome != "not_observed" {
		return ErrInvalidInput
	}
	switch o.State {
	case "draft":
		if !o.Draft {
			return ErrInvalidInput
		}
	case "open", "closed":
	case "merged":
		if !IsLowerHex(o.MergeCommit, 40) {
			return ErrInvalidInput
		}
	default:
		return ErrInvalidInput
	}
	if o.Deployment == "not_observed" && len(o.Deployments) > 0 || o.Deployment == "observed" && (len(o.Deployments) == 0 || o.DeploymentsTruncated) || o.Deployment == "truncated" && !o.DeploymentsTruncated {
		return ErrInvalidInput
	}
	switch o.Deployment {
	case "not_observed", "observed", "truncated", "unavailable":
	default:
		return ErrInvalidInput
	}
	for _, deployment := range o.Deployments {
		if deployment.ID == "" || len(deployment.ID) > 128 || deployment.RepositoryID != d.RepositoryID || deployment.ProviderProfile != d.Target.Profile || deployment.Environment == "" || len(deployment.Environment) > 256 || len(deployment.ProviderState) > 64 || deployment.ProviderUpdatedAt.IsZero() || deployment.ObservedAt.IsZero() {
			return ErrInvalidInput
		}
		if !(deployment.CommitRelation == "published_head" && deployment.Commit == o.Commit || deployment.CommitRelation == "provider_merge" && o.State == "merged" && deployment.Commit == o.MergeCommit) {
			return ErrInvalidInput
		}
		switch deployment.State {
		case "success", "failed", "pending", "inactive", "unknown":
		default:
			return ErrInvalidInput
		}
		if deployment.StatusTruncated && deployment.State != "unknown" {
			return ErrInvalidInput
		}
	}
	switch o.ChecksState {
	case "unknown", "passed", "failed", "pending", "unavailable":
	default:
		return ErrInvalidInput
	}
	if o.ChecksState == "passed" && (len(o.Checks) == 0 || o.ChecksTruncated) {
		return ErrInvalidInput
	}
	for _, check := range o.Checks {
		if check.Commit != o.Commit || len(check.ID) > 256 || len(check.Name) > 512 {
			return ErrInvalidInput
		}
		switch check.State {
		case "passed", "failed", "pending", "unexecuted", "unknown":
		default:
			return ErrInvalidInput
		}
		if o.ChecksState == "passed" && check.State != "passed" {
			return ErrInvalidInput
		}
	}
	return nil
}

// DeliveryArtifact carries exact persisted implementation evidence for human inspection.
type DeliveryArtifact struct {
	DeliveryID     string          `json:"deliveryId"`
	DeliveryDigest string          `json:"deliveryDigest"`
	ArtifactDigest string          `json:"artifactDigest"`
	Artifact       json.RawMessage `json:"artifact"`
}

// DeliveryProposalDigest excludes later authorization, receipts, observations and
// database timestamps. Its explicit shape keeps future presentation fields from
// silently changing the meaning of a person's exact publication authorization.
func DeliveryProposalDigest(d Delivery) (string, error) {
	return JSONDigest(struct {
		ID           string         `json:"id"`
		WorkspaceID  string         `json:"workspaceId"`
		RepositoryID string         `json:"repositoryId"`
		ProposerID   string         `json:"proposerId"`
		Input        DeliveryInput  `json:"input"`
		Target       DeliveryTarget `json:"target"`
		BaseCommit   string         `json:"baseCommit"`
		BaseTree     string         `json:"baseTree"`
		ResultTree   string         `json:"resultTree"`
		PatchDigest  string         `json:"patchDigest"`
		Branch       string         `json:"branch"`
	}{d.ID, d.WorkspaceID, d.RepositoryID, d.ProposerID, d.Input, d.Target, d.BaseCommit, d.BaseTree, d.ResultTree, d.PatchDigest, d.Branch})
}
